### PR TITLE
feat: Use `/ivp/pdm/device_data` to obtain detailed device statistics

### PR DESCRIPTION
**Pull request by Bidski**
This is the same as #280 but is rebased on top of #277

According to https://github.com/Matthew1471/Enphase-API/issues/29 the /ivp/pdm/device_data endpoint can provide extra details about each inverter. Specifically, it can provide panel DC voltage and current, as well as inverter AC voltage, current, and frequency.

This PR optionally uses this endpoint to get inverter data if it is available.

### CODE REVIEW

File: src/pyenphase/models/inverter.py [hunk 1]
```diff
@@ -24,3 +30,20 @@ def from_v1_api(cls, data: dict[str, Any]) -> EnvoyInverter:
...
+        """Initialize from device data."""
+        return cls(
+            serial_number=data["sn"],
+            last_report_date=data["channels"][0]["lastReading"]["endDate"],
+            last_report_watts=data["channels"][0]["watts"]["now"],
+            max_report_watts=data["channels"][0]["watts"]["max"],
+            dc_voltage=data["channels"][0]["lastReading"]["dcVoltageINmV"] / 1000.0,
+            dc_current=data["channels"][0]["lastReading"]["dcCurrentINmA"] / 1000.0,
+            ac_voltage=data["channels"][0]["lastReading"]["acVoltageINmV"] / 1000.0,
+            ac_current=data["channels"][0]["lastReading"]["acCurrentInmA"] / 1000.0,
```
Comment (catsmanac) [inline src/pyenphase/models/inverter.py:45]: Are you seeing actual values for AC current? Checked with my IQ7A inverters and reported value is zero.

Comment (Bidski) [reply src/pyenphase/models/inverter.py:45]: The test data I added is from my system

**Comment by Bidski** [comment]
How do you typically test the `EnvoyAuthenticationRequired` exception for endpoints?

**Comment by catsmanac** [comment]
> How do you typically test the `EnvoyAuthenticationRequired` exception for endpoints?

That will happen if you try to get data before authentication happens. Test_auth.py has [this](https://github.com/pyenphase/pyenphase/blob/7c69d48752e87e9f08722ffe1c3c3a2419689fad/tests/test_auth.py#L31).

If you want to try this for a specific endpoint replace the envoy.update in that example by envoy.request.

Or use a 404 response in a mcok

```py
    mock_aioresponse.get(
        "https://127.0.0.1/api/v1/production",
        status=401,
        payload={"message": "Test authentication required"},
    )
```

**Comment by catsmanac** [comment]
Ah, you meant for the probe code, then see https://github.com/pyenphase/pyenphase/blob/7c69d48752e87e9f08722ffe1c3c3a2419689fad/tests/test_auth.py#L99

**Comment by catsmanac** [comment]
😄 

```
: 2025-06-08 09:23:28 44 W dc: 34.476 V 1.769 A ac: 233.512 V 0.0 A  50.002 Hz temp 18 C energy: 0.012345016690510252 Wh today: 78 Wh lifetime: 785854.9047222222 Wh
: 2025-06-08 09:39:00 45 W dc: 32.788 V 0.736 A ac: 231.536 V 0.0 A  50.036 Hz temp 18 C energy: 0.012402241297091082 Wh today: 90 Wh lifetime: 785866.4636111112 Wh
: 2025-06-08 09:54:02 49 W dc: 34.84 V 2.002 A ac:230.952 V 0.0 A  50.026 Hz temp 18 C energy: 0.013590169992609017 Wh today: 102 Wh lifetime: 785878.7219444445 Wh
: 2025-06-08 10:09:04 83 W dc: 35.153 V 2.618 A ac: 233.728 V 0.0 A  50.034 Hz temp 19 C energy: 0.02292220990391722 Wh today: 123 Wh lifetime: 785899.3977777777 Wh
```

Looking at the data it seems my proposal to include energy produced over the collection cycle is maybe less useful. Maybe make it mWh with 3 decimals. For the value to be useful the ["lastReadings"]["duration"] value might be worthwhile to add.

Lifetime value can be an integer I think.

The lifetime energy reading confirms the quality of the HA Riemann integrator. This panel is reading 785 kWh and the Riemann I have in HA reports 782 kWh over a period of 2 years and 3 months. 0.4% difference is not bad at all. This also confirms the power reported is the average over the reporting period.

**Comment by Bidski** [comment]
> 😄
> 
> ```
> : 2025-06-08 09:23:28 44 W dc: 34.476 V 1.769 A ac: 233.512 V 0.0 A  50.002 Hz temp 18 C energy: 0.012345016690510252 Wh today: 78 Wh lifetime: 785854.9047222222 Wh
> : 2025-06-08 09:39:00 45 W dc: 32.788 V 0.736 A ac: 231.536 V 0.0 A  50.036 Hz temp 18 C energy: 0.012402241297091082 Wh today: 90 Wh lifetime: 785866.4636111112 Wh
> : 2025-06-08 09:54:02 49 W dc: 34.84 V 2.002 A ac:230.952 V 0.0 A  50.026 Hz temp 18 C energy: 0.013590169992609017 Wh today: 102 Wh lifetime: 785878.7219444445 Wh
> : 2025-06-08 10:09:04 83 W dc: 35.153 V 2.618 A ac: 233.728 V 0.0 A  50.034 Hz temp 19 C energy: 0.02292220990391722 Wh today: 123 Wh lifetime: 785899.3977777777 Wh
> ```
> 
> Looking at the data it seems my proposal to include energy produced over the collection cycle is maybe less useful. Maybe make it mWh with 3 decimals. For the value to be useful the ["lastReadings"]["duration"] value might be worthwhile to add.
> 
> Lifetime value can be an integer I think.
> 
> The lifetime energy reading confirms the quality of the HA Riemann integrator. This panel is reading 785 kWh and the Riemann I have in HA reports 782 kWh over a period of 2 years and 3 months. 0.4% difference is not bad at all. This also confirms the power reported is the average over the reporting period.

Some live data from my system
```
2025-06-09 08:45:43 122344043197 197 W dc: 33.668 V  6.352 A ac: 245.969 V  0.853 A 50.025 Hz temp 23 C energy:  54.817 mWh (903 s) today: 125.000 Wh lifetime:    110114 Wh
2025-06-09 08:44:43 122344044368 164 W dc: 33.660 V  5.188 A ac: 244.469 V  0.731 A 50.011 Hz temp 21 C energy:  45.473 mWh (903 s) today: 142.000 Wh lifetime:     79666 Wh
2025-06-09 08:46:13 122344044385 114 W dc: 34.059 V  3.547 A ac: 245.250 V  0.534 A 50.024 Hz temp 21 C energy:  31.769 mWh (903 s) today: 126.000 Wh lifetime:    111948 Wh
2025-06-09 08:46:44 122344044392  10 W dc: 30.797 V  0.445 A ac: 245.188 V  0.290 A 50.013 Hz temp 17 C energy:   2.765 mWh (904 s) today:  12.000 Wh lifetime:     99589 Wh
2025-06-09 08:50:44 122344044656  10 W dc: 30.129 V  0.477 A ac: 244.250 V  0.250 A 50.037 Hz temp 17 C energy:   2.699 mWh (903 s) today:  10.000 Wh lifetime:     76343 Wh
2025-06-09 08:50:15 122344046030  10 W dc: 30.180 V  0.516 A ac: 244.531 V  0.249 A 49.993 Hz temp 17 C energy:   2.835 mWh (904 s) today:  10.000 Wh lifetime:     85680 Wh
2025-06-09 08:46:15 122344046450 169 W dc: 33.398 V  5.273 A ac: 245.219 V  0.737 A 50.027 Hz temp 22 C energy:  47.082 mWh (904 s) today: 143.000 Wh lifetime:     79816 Wh
2025-06-09 08:48:44 122344046574   7 W dc: 30.410 V  0.375 A ac: 244.594 V  0.270 A 50.000 Hz temp 17 C energy:   2.074 mWh (904 s) today:   8.000 Wh lifetime:     62498 Wh
2025-06-09 08:46:45 122344047608 123 W dc: 18.078 V  3.965 A ac: 245.125 V  0.357 A 50.016 Hz temp 21 C energy:  34.154 mWh (904 s) today: 131.000 Wh lifetime:     80283 Wh
2025-06-09 08:47:13 122344047620  13 W dc: 35.672 V  0.500 A ac: 245.125 V  0.281 A 50.027 Hz temp 18 C energy:   3.668 mWh (903 s) today:  38.000 Wh lifetime:    103009 Wh
2025-06-09 08:47:15 122344047645 199 W dc: 33.633 V  6.379 A ac: 245.438 V  0.868 A 50.024 Hz temp 23 C energy:  55.198 mWh (933 s) today: 144.000 Wh lifetime:    112876 Wh
2025-06-09 08:47:15 122344047687 119 W dc: 35.625 V  3.445 A ac: 244.844 V  0.545 A 50.018 Hz temp 21 C energy:  32.978 mWh (904 s) today: 169.000 Wh lifetime:    107617 Wh
2025-06-09 08:47:44 122344047713 201 W dc: 33.145 V  6.648 A ac: 245.875 V  0.899 A 50.012 Hz temp 24 C energy:  55.928 mWh (932 s) today: 184.000 Wh lifetime:    112277 Wh
2025-06-09 08:50:45 122344047722  11 W dc: 30.133 V  0.496 A ac: 243.719 V  0.225 A 50.032 Hz temp 17 C energy:   2.976 mWh (903 s) today:  11.000 Wh lifetime:     87296 Wh
2025-06-09 08:45:44 122344048305 163 W dc: 32.750 V  4.441 A ac: 244.906 V  0.610 A 50.026 Hz temp 21 C energy:  45.404 mWh (903 s) today: 150.000 Wh lifetime:     81626 Wh
---
2025-06-09 09:00:47 122344043197 217 W dc: 33.863 V  6.930 A ac: 245.281 V  0.936 A 49.974 Hz temp 25 C energy:  60.357 mWh (904 s) today: 180.000 Wh lifetime:    110168 Wh
2025-06-09 08:59:48 122344044368 178 W dc: 33.477 V  5.648 A ac: 243.156 V  0.771 A 49.975 Hz temp 23 C energy:  49.378 mWh (905 s) today: 187.000 Wh lifetime:     79710 Wh
2025-06-09 09:01:17 122344044385 125 W dc: 34.977 V  3.922 A ac: 244.563 V  0.585 A 49.963 Hz temp 22 C energy:  34.845 mWh (904 s) today: 157.000 Wh lifetime:    111979 Wh
2025-06-09 09:01:48 122344044392  11 W dc: 31.176 V  0.516 A ac: 244.469 V  0.232 A 50.016 Hz temp 17 C energy:   3.180 mWh (904 s) today:  14.000 Wh lifetime:     99591 Wh
2025-06-09 09:05:48 122344044656  11 W dc: 30.316 V  0.508 A ac: 243.219 V  0.199 A 49.983 Hz temp 17 C energy:   2.973 mWh (904 s) today:  13.000 Wh lifetime:     76346 Wh
2025-06-09 09:05:18 122344046030  11 W dc: 29.875 V  0.391 A ac: 241.156 V  0.085 A 49.988 Hz temp 17 C energy:   3.184 mWh (903 s) today:  13.000 Wh lifetime:     85683 Wh
2025-06-09 09:01:49 122344046450 179 W dc: 33.453 V  5.707 A ac: 244.219 V  0.797 A 50.016 Hz temp 24 C energy:  49.786 mWh (934 s) today: 189.000 Wh lifetime:     79862 Wh
2025-06-09 09:03:47 122344046574   9 W dc: 30.969 V  0.422 A ac: 242.656 V  0.168 A 50.015 Hz temp 17 C energy:   2.422 mWh (903 s) today:  10.000 Wh lifetime:     62500 Wh
2025-06-09 09:01:50 122344047608  48 W dc: 35.145 V  1.020 A ac: 244.125 V  0.267 A 50.017 Hz temp 21 C energy:  13.260 mWh (905 s) today: 143.000 Wh lifetime:     80295 Wh
2025-06-09 09:02:17 122344047620  16 W dc: 34.547 V  0.855 A ac: 244.031 V  0.228 A 49.990 Hz temp 18 C energy:   4.425 mWh (904 s) today:  42.000 Wh lifetime:    103013 Wh
2025-06-09 09:02:19 122344047645 223 W dc: 33.043 V  7.336 A ac: 244.313 V  0.979 A 49.997 Hz temp 25 C energy:  61.878 mWh (904 s) today: 200.000 Wh lifetime:    112932 Wh
2025-06-09 09:02:19 122344047687 128 W dc: 31.355 V  4.309 A ac: 243.781 V  0.584 A 50.000 Hz temp 21 C energy:  35.606 mWh (904 s) today: 201.000 Wh lifetime:    107649 Wh
2025-06-09 09:02:47 122344047713 224 W dc: 33.086 V  7.285 A ac: 243.844 V  0.962 A 49.983 Hz temp 26 C energy:  62.292 mWh (903 s) today: 241.000 Wh lifetime:    112334 Wh
2025-06-09 09:05:49 122344047722  12 W dc: 31.121 V  0.547 A ac: 242.406 V  0.154 A 49.988 Hz temp 17 C energy:   3.388 mWh (904 s) today:  14.000 Wh lifetime:     87299 Wh
2025-06-09 09:00:50 122344048305 106 W dc: 25.410 V  3.348 A ac: 243.813 V  0.391 A 49.973 Hz temp 21 C energy:  29.456 mWh (906 s) today: 177.000 Wh lifetime:     81653 Wh
```

**Comment by catsmanac** [comment]
Nice, I see you are getting AC amps. I'm not. Checked the raw data and it's zero in there, so probably inverter type/fw or Envoy type/fw based. My inverters are IQ7A, Envoy is standard, not metered running 8.2.4264. Not an issue, just awareness for future questions when we add this to HA.

**Comment by Bidski** [comment]
> Nice, I see you are getting AC amps. I'm not. Checked the raw data and it's zero in there, so probably inverter type/fw or Envoy type/fw based. My inverters are IQ7A, Envoy is standard, not metered running 8.2.4264. Not an issue, just awareness for future questions when we add this to HA.

I have the IQ system controller INT (has an IQ gateway in it) firmware version 8.2.4345 and IQ8HC microinverters

Is your enphase account also registered as an installer/self-installer. I ask because in order to see the DC details in an Enphase app/website you need Enlighten Manager. I'm not sure how relevant that is to us being able to see these details via the web API though

**Comment by Bidski** [comment]
Is this good to be merged now?

**Comment by catsmanac** [comment]
>Is this good to be merged now?

> Suppose we will need more people to weigh in to figure out what combinations of inverters and firmware version are needed to get all values

Which brings the question how defensive we must be. Should we only enable this for recent firmwares [like ensemble does](https://github.com/pyenphase/pyenphase/blob/8ddcc24e031ee3a2768cf7795554714456bd7ea3/src/pyenphase/updaters/ensemble.py#L33).

Or add something to probe like

```py

        try:
            _ = {
                inverter["sn"]: EnvoyInverter.from_device_data(inverter)
                for id, inverter in inverters_data.items()
                if id not in ("deviceCount", "deviceDataLimit")
                and inverter["devName"] == "pcu"
            }
        except KeyError:
            return None
``` 

In combination with a .get in from_device_data for (only) the new values 
```py
            dc_voltage=last_reading.get("dcVoltageINmV",0) / 1000.0,
```

so we use original endpoint if minimal required current keys are missing. Tried it with a missing key in a simulator and it would bail out on a missing key.


We can't afford to put this out and have users with older fw end-up with neither of the 2.

I've asked the team if they can try with older firmware they have to see how those work.

**Comment by bdraco** [comment]
Should we add a new flag for detailed inverter data so its easier to check downstream in HA ie `DETAILED_INVERTERS = 8192  # New flag for detailed inverter data`

**Comment by catsmanac** [comment]
> Is this good to be merged now?

I've added some logic to handle message format deviations, if any, and some tests for it. Code was tested on an 3.x firmware also and did break anything.

So yes, unless you have any last issues I'll merge it in the morning.

**Comment by Bidski** [comment]
> > Is this good to be merged now?
> 
> I've added some logic to handle message format deviations, if any, and some tests for it. Code was tested on an 3.x firmware also and did break anything.
> 
> So yes, unless you have any last issues I'll merge it in the morning.

Those changes look reasonable to me, I can't see any issues on my end but I won't be able to test with my system until later today (which is probably your morning)

**Comment by Bidski** [comment]
I now realise that this is an absolutely horrid time to test something with solar, but still seems to work
```
2025-06-10 17:02:44 122344043197   0 W dc: 13.516 V  0.023 A ac: 242.906 V  0.071 A 49.998 Hz temp 20 C energy:   0.000 mWh (903 s) today: 1716.000 Wh lifetime:    113501 Wh
2025-06-10 16:47:10 122344044368   0 W dc: 20.477 V  0.078 A ac: 241.344 V  0.070 A 49.988 Hz temp 21 C energy:   0.201 mWh (931 s) today: 945.000 Wh lifetime:     81464 Wh
2025-06-10 16:48:42 122344044385   0 W dc: 16.566 V  0.066 A ac: 242.500 V  0.163 A 49.976 Hz temp 22 C energy:   0.134 mWh (933 s) today: 1841.000 Wh lifetime:    115578 Wh
2025-06-10 16:49:11 122344044392   0 W dc: 21.973 V  0.063 A ac: 242.531 V  0.070 A 49.973 Hz temp 21 C energy:   0.201 mWh (932 s) today: 920.000 Wh lifetime:    101425 Wh
2025-06-10 16:52:42 122344044656   0 W dc: 21.973 V  0.055 A ac: 242.469 V  0.076 A 50.001 Hz temp 21 C energy:   0.067 mWh (932 s) today: 321.000 Wh lifetime:     77029 Wh
2025-06-10 16:53:12 122344046030   0 W dc: 22.043 V  0.043 A ac: 242.844 V  0.077 A 50.017 Hz temp 21 C energy:   0.067 mWh (931 s) today: 1001.000 Wh lifetime:     87682 Wh
2025-06-10 16:49:41 122344046450   0 W dc: 22.051 V  0.059 A ac: 242.875 V  0.069 A 49.964 Hz temp 21 C energy:   0.134 mWh (931 s) today: 940.000 Wh lifetime:     81565 Wh
2025-06-10 16:51:11 122344046574   0 W dc: 22.027 V  0.047 A ac: 242.031 V  0.074 A 50.002 Hz temp 21 C energy:   0.067 mWh (931 s) today:  83.000 Wh lifetime:     62692 Wh
2025-06-10 16:50:41 122344047608   0 W dc: 21.996 V  0.066 A ac: 242.625 V  0.074 A 49.975 Hz temp 21 C energy:   0.067 mWh (930 s) today: 891.000 Wh lifetime:     81953 Wh
2025-06-10 16:53:42 122344047620   0 W dc: 21.969 V  0.031 A ac: 242.469 V  0.070 A 49.999 Hz temp 21 C energy:   0.067 mWh (930 s) today: 1041.000 Wh lifetime:    105059 Wh
2025-06-10 16:54:12 122344047645   0 W dc: 22.031 V  0.055 A ac: 242.531 V  0.070 A 49.989 Hz temp 21 C energy:   0.000 mWh (930 s) today: 1938.000 Wh lifetime:    116600 Wh
2025-06-10 16:54:42 122344047687   0 W dc: 21.984 V  0.027 A ac: 241.938 V  0.075 A 49.959 Hz temp 21 C energy:   0.000 mWh (930 s) today: 1497.000 Wh lifetime:    110446 Wh
2025-06-10 16:50:11 122344047713   0 W dc: 21.961 V  0.051 A ac: 242.656 V  0.075 A 49.982 Hz temp 21 C energy:   0.067 mWh (930 s) today: 1962.000 Wh lifetime:    116010 Wh
2025-06-10 16:54:43 122344047722   0 W dc: 21.961 V  0.035 A ac: 242.219 V  0.074 A 49.967 Hz temp 21 C energy:   0.067 mWh (930 s) today: 1082.000 Wh lifetime:     89462 Wh
2025-06-10 16:55:12 122344048305   0 W dc: 21.984 V  0.020 A ac: 241.531 V  0.074 A 50.013 Hz temp 20 C energy:   0.067 mWh (930 s) today: 959.000 Wh lifetime:     83414 Wh
```2025-06-10 17:02:44 122344043197   0 W dc: 13.516 V  0.023 A ac: 242.906 V  0.071 A 49.998 Hz temp 20 C energy:   0.000 mWh (903 s) today: 1716.000 Wh lifetime:    113501 Wh
2025-06-10 16:47:10 122344044368   0 W dc: 20.477 V  0.078 A ac: 241.344 V  0.070 A 49.988 Hz temp 21 C energy:   0.201 mWh (931 s) today: 945.000 Wh lifetime:     81464 Wh
2025-06-10 16:48:42 122344044385   0 W dc: 16.566 V  0.066 A ac: 242.500 V  0.163 A 49.976 Hz temp 22 C energy:   0.134 mWh (933 s) today: 1841.000 Wh lifetime:    115578 Wh
2025-06-10 16:49:11 122344044392   0 W dc: 21.973 V  0.063 A ac: 242.531 V  0.070 A 49.973 Hz temp 21 C energy:   0.201 mWh (932 s) today: 920.000 Wh lifetime:    101425 Wh
2025-06-10 16:52:42 122344044656   0 W dc: 21.973 V  0.055 A ac: 242.469 V  0.076 A 50.001 Hz temp 21 C energy:   0.067 mWh (932 s) today: 321.000 Wh lifetime:     77029 Wh
2025-06-10 16:53:12 122344046030   0 W dc: 22.043 V  0.043 A ac: 242.844 V  0.077 A 50.017 Hz temp 21 C energy:   0.067 mWh (931 s) today: 1001.000 Wh lifetime:     87682 Wh
2025-06-10 16:49:41 122344046450   0 W dc: 22.051 V  0.059 A ac: 242.875 V  0.069 A 49.964 Hz temp 21 C energy:   0.134 mWh (931 s) today: 940.000 Wh lifetime:     81565 Wh
2025-06-10 16:51:11 122344046574   0 W dc: 22.027 V  0.047 A ac: 242.031 V  0.074 A 50.002 Hz temp 21 C energy:   0.067 mWh (931 s) today:  83.000 Wh lifetime:     62692 Wh
2025-06-10 16:50:41 122344047608   0 W dc: 21.996 V  0.066 A ac: 242.625 V  0.074 A 49.975 Hz temp 21 C energy:   0.067 mWh (930 s) today: 891.000 Wh lifetime:     81953 Wh
2025-06-10 16:53:42 122344047620   0 W dc: 21.969 V  0.031 A ac: 242.469 V  0.070 A 49.999 Hz temp 21 C energy:   0.067 mWh (930 s) today: 1041.000 Wh lifetime:    105059 Wh
2025-06-10 16:54:12 122344047645   0 W dc: 22.031 V  0.055 A ac: 242.531 V  0.070 A 49.989 Hz temp 21 C energy:   0.000 mWh (930 s) today: 1938.000 Wh lifetime:    116600 Wh
2025-06-10 16:54:42 122344047687   0 W dc: 21.984 V  0.027 A ac: 241.938 V  0.075 A 49.959 Hz temp 21 C energy:   0.000 mWh (930 s) today: 1497.000 Wh lifetime:    110446 Wh
2025-06-10 16:50:11 122344047713   0 W dc: 21.961 V  0.051 A ac: 242.656 V  0.075 A 49.982 Hz temp 21 C energy:   0.067 mWh (930 s) today: 1962.000 Wh lifetime:    116010 Wh
2025-06-10 16:54:43 122344047722   0 W dc: 21.961 V  0.035 A ac: 242.219 V  0.074 A 49.967 Hz temp 21 C energy:   0.067 mWh (930 s) today: 1082.000 Wh lifetime:     89462 Wh
2025-06-10 16:55:12 122344048305   0 W dc: 21.984 V  0.020 A ac: 241.531 V  0.074 A 50.013 Hz temp 20 C energy:   0.067 mWh (930 s) today: 959.000 Wh lifetime:     83414 Wh

**Comment by Bidski** [comment]
> for solar data, yes, on the other hand though, good time to see if zero values are dealt with properly.

Agreed, the numbers look as I would expect them to, given time of day, etc

**Comment by catsmanac** [comment]
>  Will you guys be updating with HA or would you like me to make an attempt at that?

Whatever you like. I have some initial preps ready so I can go ahead create a draft one. But if you want to give it a try, feel free to go ahead and we can review as in this one.

**Comment by Bidski** [comment]
> Whatever you like. I have some initial preps ready so I can go ahead create a draft one. But if you want to give it a try, feel free to go ahead and we can review as in this one. 

This will be my first foray into HA dev, so you will definitely be a lot faster than me, but if you have better things to do in the short term then I am happy to give it a crack

**Comment by catsmanac** [comment]
Give it a crack I'd say, good learning exercise (That's how I got going with help from the others). In HA I'm a mere mortal (as it should) and can't finally approve and merge PR. So I can add to your commits as needed, and the other team members can do the final review and merge.

We have time as this will be a new feature which are only included in major updates, so 2025.7 at best.

Things to do:

- bump pyenphase dependancy to 2.1.0 by updating manifest.py
- then run: python3 -m script.gen_requirements_all  to generate requirements files.
- You now have 3 updated files for the bump
- Then in sensors.py add the new ones to `INVERTER_SENSORS = (`
  something like
```py
    EnvoyInverterSensorEntityDescription(
        key="dc_voltage",
        translation_key="dc_voltage",
        native_unit_of_measurement=UnitOfElectricPotential.VOLT,
        state_class=SensorStateClass.MEASUREMENT,
        device_class=SensorDeviceClass.VOLTAGE,
        suggested_display_precision=3,
        entity_registry_enabled_default=False,
        value_fn=attrgetter("dc_voltage"),
    ),
```
- Note that the example adds the new ones in disabled state. This to avoid sudden increase in entity and statistics size. Let's the user make the choise to enable or not.
- Add translations for the new sensors to strings.json. something like:
```json
      "dc_voltage": {
        "name": "DC voltage"
      },
```
- update diagnostics.py to add the new endpoint
- test and fixtures need to test with all None and actual values for the inverters. Fixtures need the new fields and at least one should have none values and at least one should have actual values.
- Documentation PR as well. I'll start that one (they need to reference each other).
- mark the PR as new feature and dependancy upgrade and work in progress.

**Comment by Bidski** [comment]
Still working on the unit tests, but otherwise I think this is working https://github.com/home-assistant/core/pull/146916
